src/inpod/protocol.rs

```diff
@@ -0,0 +1,409 @@
+            match cmsg {
+                ControlMessageOwned::ScmRights(fds) => {
+                    let len = fds.len();
+                    total_fds += len;
+                    if total_fds != 1 {
+                        for fd in fds {
+                            // fds in the vector are ours, so own them and drop them so they are closed (prevent resource leak).
+                            // Safety: ScmRights returns a list of FDs that we own, so we can safely drop them.
+                            std::mem::drop(unsafe { std::os::fd::OwnedFd::from_raw_fd(fd) });
+                        }
+                    } else {
+                        // Safety: ScmRights returns FDs opened by the kernel for us, so we can
+                        // safely own it.
+                        our_netns = Some(unsafe { std::os::fd::OwnedFd::from_raw_fd(fds[0]) })
+                    }
+                }
+                u => {
+                    warn!("Unexpected control message {:?}", u);
+                    continue;
+                }
+            }
+        }
+        // only check for errors once we are done parsing all FDs
+        if total_fds > 1 {
+            anyhow::bail!("Expected 1 FD, got {}", total_fds);
+        }
+
+        // make sure that we got a netns FD.
+        if let Some(our_netns) = &our_netns {
+            // validate that the fd we got is a netns. This should never happen, and is here
+            // to catch potential bugs in the node agent during development.
+            debug!(
+                "Validating netns FD: {:?}",
+                validate_ns(our_netns.as_raw_fd())
+            );
+        }
+
+        Ok(our_netns)
+    }
+}
+
+fn validate_ns(fd: std::os::fd::RawFd) -> anyhow::Result<()> {
+    // on newer kernels we can get the ns type! note that this doesn't work on older kernels.
+    // so an error doesn't mean its not a netns.
+    // #define NSIO	0xb7
+    const NSIO: u8 = 0xb7;
+    // #define NS_GET_NSTYPE		_IO(NSIO, 0x3)
+    const NS_GET_NSTYPE: u8 = 0x3;
+    nix::ioctl_none!(get_ns_type, NSIO, NS_GET_NSTYPE);
+    let nstype = unsafe { get_ns_type(fd) };
+    if let Ok(nstype) = nstype {
+        // ignore errors in case we are in an old kernel
+        if nstype != nix::libc::CLONE_NEWNET {
+            anyhow::bail!("Unexpected ns type: {:?}", nstype);
+        } else {
+            debug!("FD {} type is netns", fd);
+        }
+    } else {
+        // can get ns type, do a different check - that the fd came from the nsfs.
+        let data = nix::sys::statfs::fstatfs(&fd)?;
+        let f_type = data.filesystem_type();
+        if f_type != nix::sys::statfs::PROC_SUPER_MAGIC && f_type != nix::sys::statfs::NSFS_MAGIC {
+            anyhow::bail!("Unexpected FD type for netns: {:?}", f_type);
+        }
+    }
+
+    debug!("FD {} looks like a netns", fd);
+    Ok(())
+}
+
+fn iovs_buf<'a>(slices: &'a [IoSliceMut<'a>], remaining: usize) -> IoSliceBuf<'a> {
```

Comment:
Why do we need all of this custom implementation? Isn't there always only a single element in the slices list? Seems like we can just just directly translate to bytes::Bytes from that and remove all of this code. 

or if we really need multiple buffers for some reason,

```rust
        bytes::Bytes::from_iter(slices
            .iter()
            .flat_map(|slice| slice.as_ref())
            .map(|x| *x))
```

Reply:
I might be able to do that. it will require a bit restructuring to out-smart the borrow checker.

The original reason is that i couldn't pass `&[u8]` to `get_workload_data` as it was already borrowed; so i had to pass the `iov`.  I didn't want the code to look weird by me just referring only to `iov[0], hence `iovs_buf`.